jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSourcePublisher.java

```diff
@@ -13,147 +13,318 @@
 
 package org.eclipse.jetty.io.content;
 
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CompletionException;
 import java.util.concurrent.Flow;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
 
 import org.eclipse.jetty.io.Content;
+import org.eclipse.jetty.util.IteratingCallback;
 import org.eclipse.jetty.util.MathUtils;
-import org.eclipse.jetty.util.thread.AutoLock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * <p>Wraps a {@link Content.Source} as a {@link Flow.Publisher}.
  * When content is requested via {@link Flow.Subscription#request(long)}, it is
  * read from the passed {@link Content.Source} and passed to {@link Flow.Subscriber#onNext(Object)}.
  * If no content is available, then the {@link Content.Source#demand(Runnable)} method is used to
  * ultimately call {@link Flow.Subscriber#onNext(Object)} once content is available.</p>
+ * <p>{@link Content.Source} can be consumed only once and does not support multicast subscription.
+ * {@link Content.Source} will be consumed fully, otherwise will be failed in case of any errors
+ * to prevent resource leaks.</p>
  */
 public class ContentSourcePublisher implements Flow.Publisher<Content.Chunk>
 {
-    private final Content.Source content;
+    private static final Logger LOG = LoggerFactory.getLogger(ContentSourcePublisher.class);
+
+    private final AtomicBoolean exhausted;
+    private Content.Source content;
 
     public ContentSourcePublisher(Content.Source content)
     {
+        if (content == null)
+            throw new IllegalArgumentException("Content.Source must not be null");
+        this.exhausted = new AtomicBoolean(false);
         this.content = content;
     }
 
     @Override
     public void subscribe(Flow.Subscriber<? super Content.Chunk> subscriber)
     {
-        subscriber.onSubscribe(new SubscriptionImpl(content, subscriber));
+        // As per rule 1.11, we have decided to support SINGLE subscriber
+        // in a UNICAST configuration for this implementation. It means
+        // that Content.Source can be consumed only once.
+        if (this.exhausted.compareAndSet(false, true))
+            subscribeActive(subscriber);
+        else
+            subscribeExhausted(subscriber);
     }
 
-    private static class SubscriptionImpl implements Flow.Subscription
+    private void subscribeActive(Flow.Subscriber<? super Content.Chunk> subscriber)
```

Comment:
A name that would better match our naming conventions would be:
```suggestion
    private void onSubscribe(Flow.Subscriber<? super Content.Chunk> subscriber)
```

Reply:
thanks for the suggestion, fixed